apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.PostPolicyElements.cs

```diff
@@ -0,0 +1,459 @@
+        /// Allowed values are 200, 201 and 204.
+        /// </summary>
+        internal sealed class SuccessActionStatus : ISupportsExact<HttpStatusCode>
+        {
+            /// <summary>
+            /// The default success-action-status element.
+            /// </summary>
+            public static SuccessActionStatus Element { get; } = new SuccessActionStatus();
+
+            internal SuccessActionStatus() { }
+
+            /// <inheritdoc/>
+            string IPostPolicyElement.ElementName => "success-action-status";
+
+            /// <inheritdoc/>
+            object ISupportsExact<HttpStatusCode>.ToPostPolicyValue(HttpStatusCode value) =>
+                ((int)value).ToString();
+        }
+
+        /// <summary>
+        /// Custom element used for requiering extra Google metadata to be included
+        /// in the post request.
+        /// </summary>
+        /// <typeparam name="TValue">The type for the values of the element.</typeparam>
+        internal sealed class GoogleMetadata<TValue> : ISupportsExact<TValue>, ISupportsStartsWith
+        {
+            private const string s_elementPrefix = "x-goog-meta-";
+
+            private readonly Func<TValue, object> _valueToObjectFunc;
+            private readonly string _elementName;
+
+            /// <summary>
+            /// Creates a new element with the given name.
+            /// If the given name is not prefixed with x-goog-meta- it will be so.
+            /// </summary>
+            /// <param name="name">The name for this element that will be lowercased and
+            /// prefixed with x-goog-meta- (if not already so) to obtain the element name. Must not be null.</param>
+            /// <param name="valueToObjectFunc">The function to use to convert <typeparamref name="TValue"/> values
+            /// to the representation with which they'll be included in a <see cref="PostPolicy"/>.
+            /// May be null in which case <typeparamref name="TValue"/> ToString() will be used.</param>
+            internal GoogleMetadata(string name, Func<TValue, object> valueToObjectFunc = null)
+            {
+                _elementName = GaxPreconditions.CheckNotNull(name, nameof(name)).ToLowerInvariant();
+                if (!_elementName.StartsWith(s_elementPrefix))
+                {
+                    _elementName = $"{s_elementPrefix}{_elementName}";
+                }
+                _valueToObjectFunc = valueToObjectFunc ?? (value => value.ToString());
+            }
+
+            /// <inheritdoc/>
+            string IPostPolicyElement.ElementName => _elementName;
+
+            /// <inheritdoc/>
+            object ISupportsExact<TValue>.ToPostPolicyValue(TValue value) =>
+                _valueToObjectFunc(value);
+
+            /// <inheritdoc/>
+            public override bool Equals(object obj) =>
+                obj is GoogleMetadata<TValue> other ? _elementName.Equals(other._elementName) : false;
```

Comment:
I'd write this as `obj is GoogleMetadata<TValue> other && _elementName == other.elementName`